NPE in org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitMemberDeclarator

After merging the latest updates the following NPE is thrown:

```
ERROR: Error during Sonar runner execution
org.sonar.runner.impl.RunnerException: Unable to execute Sonar
        at org.sonar.runner.impl.BatchLauncher$1.delegateExecution(BatchLauncher.java:91)
        at org.sonar.runner.impl.BatchLauncher$1.run(BatchLauncher.java:75)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.sonar.runner.impl.BatchLauncher.doExecute(BatchLauncher.java:69)
        at org.sonar.runner.impl.BatchLauncher.execute(BatchLauncher.java:50)
        at org.sonar.runner.api.EmbeddedRunner.doExecute(EmbeddedRunner.java:102)
        at org.sonar.runner.api.Runner.execute(Runner.java:100)
        at org.sonar.runner.Main.executeTask(Main.java:70)
        at org.sonar.runner.Main.execute(Main.java:59)
        at org.sonar.runner.Main.main(Main.java:53)
Caused by: org.sonar.squid.api.AnalysisException: Unable to analyze file: C:\TFS_Server\Server\xxxxxxxxxxxxxxxxxxxxxxxxxxxxx.h
        at com.sonar.sslr.squid.AstScanner.scanFiles(AstScanner.java:123)
        at org.sonar.plugins.cxx.squid.CxxSquidSensor.analyse(CxxSquidSensor.java:104)
        at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:79)
        at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:70)
        at org.sonar.batch.phases.PhaseExecutor.execute(PhaseExecutor.java:119)
        at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:194)
        at org.sonar.api.platform.ComponentContainer.startComponents(ComponentContainer.java:92)
        at org.sonar.api.platform.ComponentContainer.execute(ComponentContainer.java:77)
        at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:233)
        at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:228)
        at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:221)
        at org.sonar.api.platform.ComponentContainer.startComponents(ComponentContainer.java:92)
        at org.sonar.api.platform.ComponentContainer.execute(ComponentContainer.java:77)
        at org.sonar.batch.scan.ScanTask.scan(ScanTask.java:64)
        at org.sonar.batch.scan.ScanTask.execute(ScanTask.java:51)
        at org.sonar.batch.bootstrap.TaskContainer.doAfterStart(TaskContainer.java:125)
        at org.sonar.api.platform.ComponentContainer.startComponents(ComponentContainer.java:92)
        at org.sonar.api.platform.ComponentContainer.execute(ComponentContainer.java:77)
        at org.sonar.batch.bootstrap.BootstrapContainer.executeTask(BootstrapContainer.java:173)
        at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:95)
        at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:67)
        at org.sonar.runner.batch.IsolatedLauncher.execute(IsolatedLauncher.java:48)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.sonar.runner.impl.BatchLauncher$1.delegateExecution(BatchLauncher.java:87)
        ... 9 more
Caused by: java.lang.NullPointerException
        at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitMemberDeclarator(AbstractCxxPublicApiVisitor.java:317)
        at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitMemberDeclaration(AbstractCxxPublicApiVisitor.java:269)
        at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitNode(AbstractCxxPublicApiVisitor.java:141)
        at com.sonar.sslr.impl.ast.AstWalker.visitNode(AstWalker.java:114)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:85)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
        at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
        at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:69)
        at com.sonar.sslr.squid.AstScanner.scanFiles(AstScanner.java:102)
        ... 35 more
```

--- a/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
@@ -305,19 +305,39 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
             comments = getDeclaratorInlineComment(node);
         }
 
-        AstNode idNode = node.getFirstDescendant(CxxGrammarImpl.declaratorId);
-        String id;
+        // find the identifier to present to concrete visitors
+        String id = null;
+        AstNode idNode = null;
 
-        AstNode operatorFunctionId = node
-                .getFirstDescendant(CxxGrammarImpl.operatorFunctionId);
+        // first look for an operator function id
+        idNode = node.getFirstDescendant(CxxGrammarImpl.operatorFunctionId);
 
-        if (operatorFunctionId != null) {
-            id = getOperatorId(operatorFunctionId);
-        } else {
-            id = idNode.getTokenValue();
+        if (idNode != null) {
+            id = getOperatorId(idNode);
+        }
+        else {
+            // look for a declarator id
+            idNode = node.getFirstDescendant(CxxGrammarImpl.declaratorId);
+
+            if (idNode != null) {
+                id = idNode.getTokenValue();
+            }
+            else {
+                // look for an identifier (e.g in bitfield declaration)
+                idNode = node.getFirstDescendant(GenericTokenType.IDENTIFIER);
+
+                if (idNode != null) {
+                    id = idNode.getTokenValue();
+                }
+                else {
+                    LOG.error("Unsupported declarator at " + node.getTokenLine());
+                }
+            }
         }
 
-        visitPublicApi(idNode, id, comments);
+        if (idNode != null && id != null) {
+            visitPublicApi(idNode, id, comments);
+        }
     }
 
     private void visitEnumSpecifier(AstNode enumSpecifierNode) {

--- a/cxx-squid/src/test/java/org/sonar/cxx/CxxPublicApiVisitorTest.java
+++ b/cxx-squid/src/test/java/org/sonar/cxx/CxxPublicApiVisitorTest.java
@@ -123,7 +123,7 @@ public class CxxPublicApiVisitorTest {
 
     @Test
     public void to_delete() {
-        testFile("src/test/resources/metrics/public_api.h", 34, 0, true);
+        testFile("src/test/resources/metrics/public_api.h", 35, 0, true);
     }
 
     @Test
@@ -205,6 +205,7 @@ public class CxxPublicApiVisitorTest {
                 "protectedStructField2");
         expectedIdCommentMap.put("protectedClass", "protectedClass");
         expectedIdCommentMap.put("operator[]", "operator");
+        expectedIdCommentMap.put("bitfield", "bitfield");
 
         // check completeness
         for (final String id : expectedIdCommentMap.keySet()) {

--- a/cxx-squid/src/test/resources/metrics/public_api.h
+++ b/cxx-squid/src/test/resources/metrics/public_api.h
@@ -87,6 +87,11 @@ struct testStruct {
 	/// testTypeDef
 	///
     typedef toto<T> testTypeDef;
+
+    /**
+     * bitfield doc
+     */
+    unsigned int bitfield:1;
 };
 
 /**

